hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/common/AbstractNativeSystemContract.java

```diff
@@ -0,0 +1,189 @@
+import static com.hedera.node.app.service.contract.impl.utils.SystemContractUtils.successResultOf;
+import static com.hedera.node.app.service.evm.utils.ValidationUtils.validateTrue;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_TRANSACTION_BODY;
+import static java.util.Objects.requireNonNull;
+import static org.hyperledger.besu.evm.frame.ExceptionalHaltReason.INVALID_OPERATION;
+import static org.hyperledger.besu.evm.frame.ExceptionalHaltReason.PRECOMPILE_ERROR;
+
+import com.hedera.hapi.node.base.ContractID;
+import com.hedera.hapi.node.base.ResponseCodeEnum;
+import com.hedera.node.app.service.contract.impl.exec.failure.CustomExceptionalHaltReason;
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.AbstractFullContract;
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.FullResult;
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.HederaSystemContract;
+import com.hedera.node.app.service.contract.impl.exec.utils.FrameUtils;
+import com.hedera.node.app.service.contract.impl.hevm.HederaWorldUpdater;
+import com.hedera.node.app.spi.workflows.HandleException;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import javax.inject.Singleton;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+import org.hyperledger.besu.evm.gascalculator.GasCalculator;
+
+/**
+ * Abstract class for native system contracts.
+ * Descendents are {@link com.hedera.node.app.service.contract.impl.exec.systemcontracts.HtsSystemContract} and
+ * {@link com.hedera.node.app.service.contract.impl.exec.systemcontracts.HasSystemContract}.
+ */
+@Singleton
+public abstract class AbstractNativeSystemContract extends AbstractFullContract implements HederaSystemContract {
+    private static final Logger log = LogManager.getLogger(AbstractNativeSystemContract.class);
+    private final CallFactory callFactory;
+    private final ContractID contractID;
+
+    protected AbstractNativeSystemContract(
+            @NonNull String name,
+            @NonNull CallFactory callFactory,
+            @NonNull ContractID contractID,
+            @NonNull GasCalculator gasCalculator) {
+        super(name, gasCalculator);
+        this.callFactory = callFactory;
+        this.contractID = contractID;
+    }
+
+    @Override
+    public FullResult computeFully(@NonNull final Bytes input, @NonNull final MessageFrame frame) {
+        requireNonNull(input);
+        requireNonNull(frame);
+        final var callType = callTypeOf(frame);
+        if (callType == UNQUALIFIED_DELEGATE) {
+            return haltResult(PRECOMPILE_ERROR, frame.getRemainingGas());
+        }
+        final Call call;
+        final AbstractCallAttempt attempt;
+        try {
+            validateTrue(input.size() >= 4, INVALID_TRANSACTION_BODY);
```

Comment:
Is this `4` the length of a selector, which is required? Or something else? Maybe a named constant?